kpi/models/kpi.py

```diff
@@ -0,0 +1,196 @@
+    'GRANT',
+    'REVOKE',
+    'INDEX',
+)) + ')')
+
+
+def is_sql_or_ddl_statement(query):
+    """Check if sql query is a SELECT statement"""
+    return not RE_SELECT_QUERY.match(query.upper())
+
+
+class KPI(models.Model):
+    """Key Performance Indicators."""
+
+    _name = "kpi"
+    _description = "Key Performance Indicator"
+
+    name = fields.Char('Name', required=True)
+    description = fields.Text('Description')
+    category_id = fields.Many2one(
+        'kpi.category',
+        'Category',
+        required=True,
+    )
+    threshold_id = fields.Many2one(
+        'kpi.threshold',
+        'Threshold',
+        required=True,
+    )
+    periodicity = fields.Integer('Periodicity', default=1)
+
+    periodicity_uom = fields.Selection((
+        ('hour', 'Hour'),
+        ('day', 'Day'),
+        ('week', 'Week'),
+        ('month', 'Month')
+    ), 'Periodicity UoM', required=True, default='day')
+
+    next_execution_date = fields.Datetime(
+        'Next execution date',
+        readonly=True,
+    )
+    value = fields.Float(string='Value',
+                         compute="_compute_display_last_kpi_value",
+                         )
+    color = fields.Text('Color', compute="_compute_display_last_kpi_value",)
+    last_execution = fields.Datetime(
+        'Last execution', compute="_compute_display_last_kpi_value",)
+    kpi_type = fields.Selection((
+        ('python', 'Python'),
+        ('local', 'SQL - Local DB'),
+        ('external', 'SQL - External DB')
+    ), 'KPI Computation Type')
+
+    dbsource_id = fields.Many2one(
+        'base.external.dbsource',
+        'External DB Source',
+    )
+    kpi_code = fields.Text(
+        'KPI Code',
+        help=("SQL code must return the result as 'value' "
+              "(i.e. 'SELECT 5 AS value')."),
+    )
+    history_ids = fields.One2many(
+        'kpi.history',
+        'kpi_id',
+        'History',
+    )
+    active = fields.Boolean(
+        'Active',
+        help=("Only active KPIs will be updated by the scheduler based on"
+              " the periodicity configuration."), default=True
+    )
+    company_id = fields.Many2one(
+        'res.company', 'Company',
+        default=lambda self: self.env.user.company_id.id)
+
+    @api.multi
+    def _compute_display_last_kpi_value(self):
+        history_obj = self.env['kpi.history']
+        for obj in self:
+            history_ids = history_obj.search([("kpi_id", "=", obj.id)])
+            if history_ids:
+                his = obj.history_ids[0]
+                obj.value = his.value
+                obj.color = his.color
+                obj.last_execution = his.date
+            else:
+                obj.value = 0
+                obj.color = '#FFFFFF'
+                obj.last_execution = False
+
+    @api.multi
+    def compute_kpi_value(self):
+        for obj in self:
+            kpi_value = 0
+            if obj.kpi_code:
+                if obj.kpi_type == 'local' and is_sql_or_ddl_statement(
```

Comment:
Maybe, it should be better to add a hook if someone wants to make another system of Kpis.

Reply:
The function should return the dict with value, color and kpi_id.

This way, we could use the python code to compute the color dynamically (compare with Previous year for example)

Reply:
@etobella Done!